content/courses/intro-to-embedded-sys-prog/chapters/low_level_programming.rst

```diff
@@ -391,24 +400,209 @@ Here's the generic procedure again, now showing the overlay:
+address |mdash| strictly speaking its starting address |mdash| which,
+not surprisingly, is the address of the :ada:`Buffer` component at index
+:ada:`Start`. Applying the :ada:`Address` attribute gives us that
+location.
+
+There are other characteristics we might want to query too. For example,
+we might want to ask the compiler what alignment it chose for a
+given object (or type, for all such objects). Rather than describe all
+the possibilities, we can just say that all the representation
+characteristics that can be specified can also be queried. We cover
+specifying representation characteristics next, so just assume the
+corresponding queries are available.
+
+That said, there is one particular representation query we need to talk about
+explicitly, now, because there is a lot of confusion about it: the
+:ada:`'Size` attribute. The confusion stems from the fact that there are
+multiple contexts for applying the attribute, and multiple reasonable
+interpretations possible. We can apply the :ada:`'Size` attribute to a
+type, in an attempt to get information about all objects of the type, or
+we can apply it to individual objects to get specific information. In
+both cases, what actual information do we get? In the original version
+of Ada these questions weren't really answered so vendors did what they
+thought was correct. But they did not agree with each other, and
+portability became a problem.
+
+For example, suppose you want to convert some value to a series of bytes
+in order to send the value over the wire. To do that you need to know
+how many bytes are require to represent the value. Many applications
+queried the size of the type to determine that, and then, when porting
+to a new vendor's compiler, found that their code no longer worked
+correctly. The new vendor's implementation wasn't wrong, it was just
+different.
+
+Later versions of Ada answered these questions, where possible,
+so let's examine the contexts and meaning. Above all, though, remember
+that :ada:`'Size` returns values in terms of **bits**.
+
+If we apply :ada:`'Size` to a type, the resulting value depends on the
+kind of type.
+
+For scalar types, i.e., those without components such as integers and
+enumerations, the attribute returns the *minimum* number of bits
+required to represent all the values of the type. Consider type Boolean,
+which has two possible values. One bit will suffice, and indeed the
+language standard requires Boolean'Size to be the value 1.
+
+This meaning also applies to subtypes, which can constrain the number of
+values for a scalar type. Consider subtype Natural. That's a subtype
+defined by the language to be type Integer but with a range of 0 ..
+Integer'Last. On a 32-bit machine we would expect Integer to be a native
+type, and thus 32-bits. On such a machine if we say Integer'Size we will
+indeed get 32. But if we say Natural'Size we will get 31, not 32,
+because only 31 bits are needed to represent that range on that machine.
```

Comment:
```suggestion
defined by the language to be type :ada:`Integer` but with a range of 
:ada:`0 .. Integer'Last`. 
On a 32-bit machine we would expect Integer to be a native type, and
thus 32-bits. On such a machine if we say :ada:`Integer'Size` we will
indeed get 32. But if we say :ada:`Natural'Size` we will get 31, not 32,
because only 31 bits are needed to represent that range on that machine.
```